### ISSUE TITLE
Embedding layers in dlib 

### ISSUE THREAD

**Issue by ChristopheLanuit** [davisking/dlib#1534]
Hi ! I'm using dlib for very short time and find it great. Going through the examples, the documentation & the code, I was looking for a layer class (let's call it _embedding layer_) to map discrete input data from a vocabulary to continuous (and trainable) vectors. 
 
I was wondering if I missed something (humm.. please forgive me if this is the case) or if this was not implemented at this point in time? ...in which case I think it would be a very valuable enhancement (PR submission?).

Thanks.

**Comment by ChristopheLanuit** [comment]
Thanks Davis. Agree with you, I was also wondering if the embedding processing should be before the network input (like you suggested). moreover, design-wise, a layer that takes as input discrete data and outputs continuous values is not consistent with other layers types. It's probably the solution I will try first. (Somehow it's also what I'm used with TensorFlow).
 
What made me think it could be a layer (like in Keras) is because there is a bi-directional processing (forward and backward, or preprocessing and postprocessing): to update the vectors values, I would need to get and apply gradients.

**Comment by davisking** [comment]
Why are gradients involved in this case? There aren’t any parameters to learn here right?

**Comment by ChristopheLanuit** [comment]
There are parameters to learn indeed.

Let's take the example of vector representations of words. Each word in a vocabulary is to be represented with a vector with N dimensions. Learning word embeddings from text can help identifying semantic proximity between words, reflected as a short distance between two vectors. In such training, the training data would include discrete information (word indices in the vocabulary), the embedding vectors parameters would be randomly initialized and the training process would affect the weights of the vectors. 

Two examples : the [Word2Vec Tensorflow](https://www.tensorflow.org/tutorials/representation/word2vec) example, and a [simple sentiment analysis using Keras](https://machinelearningmastery.com/use-word-embedding-layers-deep-learning-keras/). As I mentionned in my earlier post, an embedding layer in Keras can be used as an input layer. In the Keras example, the network is Embeddings->Flatten->FC(1 unit)->Sigmoid. (and a binary cross entropy loss function in this case).

**Comment by davisking** [comment]
Ah, my mistake, I thought you meant something simpler.  I'm familiar with those approaches and yes, for that you would want to define a special layer. If you want to submit a PR with such a layer that would be super cool :)
